Sources/FoundationEssentials/Subprocess/Subprocess+IO.swift

```diff
@@ -0,0 +1,267 @@
+            }
+        }
+
+        internal func closeParentSide() throws {
+            switch self {
+            case .noInput(_), .fileDescriptor(_, _):
+                break
+            case .customWrite(_, _):
+                // The parent fd should have been closed
+                // in the `body` when writer.finish() is called
+                break
+            }
+        }
+
+        internal func closeAll() throws {
+            switch self {
+            case .noInput(let readFd):
+                try readFd.close()
+            case .customWrite(let readFd, let writeFd):
+                try readFd.close()
+                try writeFd.close()
+            case .fileDescriptor(let fd, _):
+                try fd.close()
+            }
+        }
+    }
+
+    internal enum ExecutionOutput {
+        case discarded(FileDescriptor)
+        case fileDescriptor(FileDescriptor, Bool)
+        case collected(Int, FileDescriptor, FileDescriptor)
+
+        internal func getWriteFileDescriptor() -> FileDescriptor {
+            switch self {
+            case .discarded(let writeFd):
+                return writeFd
+            case .fileDescriptor(let writeFd, _):
+                return writeFd
+            case .collected(_, _, let writeFd):
+                return writeFd
+            }
+        }
+
+        internal func getReadFileDescriptor() -> FileDescriptor? {
+            switch self {
+            case .discarded(_), .fileDescriptor(_, _):
+                return nil
+            case .collected(_, let readFd, _):
+                return readFd
+            }
+        }
+
+        internal func closeChildSide() throws {
+            switch self {
+            case .discarded(let writeFd):
+                try writeFd.close()
+            case .fileDescriptor(let fd, let closeWhenDone):
+                // User passed fd
+                if closeWhenDone {
+                    try fd.close()
+                }
+                break
+            case .collected(_, _, let writeFd):
+                try writeFd.close()
+            }
+        }
+
+        internal func closeParentSide() throws {
+            switch self {
+            case .discarded(_), .fileDescriptor(_, _):
+                break
+            case .collected(_, let readFd, _):
+                try readFd.close()
+            }
+        }
+
+        internal func closeAll() throws {
+            switch self {
+            case .discarded(let writeFd):
+                try writeFd.close()
+            case .fileDescriptor(let fd, _):
+                try fd.close()
+            case .collected(_, let readFd, let writeFd):
+                try readFd.close()
+                try writeFd.close()
+            }
+        }
+    }
+}
+
+// MARK: - Private Helpers
+extension FileDescriptor {
+    internal func read(upToLength maxLength: Int) throws -> [UInt8] {
+        let buffer: UnsafeMutableBufferPointer<UInt8> = .allocate(capacity: maxLength)
```

Comment:
This buffer is never deallocated - is there a way to do this a bit safer without allocating the buffer ourself? Also, should this return a `Data` instead of a `[UInt8]`